crates/bevy_render/src/render_graph/graph.rs

```diff
@@ -92,12 +92,21 @@ impl RenderGraph {
     where
         T: Node,
     {
-        let id = NodeId::new();
         let name = name.into();
-        let mut node_state = NodeState::new(id, node);
-        node_state.name = Some(name.clone());
-        self.nodes.insert(id, node_state);
-        self.node_names.insert(name, id);
+        let id = *self
+            .node_names
+            .entry(name.clone())
+            .or_insert_with(NodeId::new);
+
+        let mut new_node_state = NodeState::new(id, node);
+        new_node_state.name = Some(name);
+        self.nodes
+            .entry(id)
+            .and_modify(|old_entry| {
+                // If replacing a node, retain the connectivity by moving them into the new node.
```

Comment:
There is no guarantee that the edges from the old node will still be valid. This could invalidate the graph.
I don't think it makes to do this in the first place:  inserting a new node for something that already exists "re-contextualizes"  the node name. Any edges built with the previous context should be discarded because the edge authors were not aware of this new context.

Honestly, I think inserting a node with the same name should be an error. Developers trying to override existing nodes should be aware that this is happening / choose how to handle this (via some combination of querying existing edges, removing the previous node with the desired name, etc).

Reply:
These are valid points, but often times we would create nodes "inherited from" or otherwise compatible with existing nodes. It should be easy to replace these nodes directly.
We can add a few different methods to cover these different use cases:
- `add_node` will always panic if the node was already added.
- `replace_node` will replace the existing node with the new one. It will also validate the edge dependencies and panic if there's a mismatch.
- `node_entry` provides an API similar to `HashMap::entry` and allows the user to specify a callback to relink the connected edges.

Please lmk if this sounds good to you.

Reply:
Sorry for the late reply. I let this get buried. I think it makes more sense to just provide the core operations. I don't think there can be a "one size fits all" replace method. To "replace a node", I think you should query the node's existing edges, remove the node, and then recreate the edges as necessary  (according to the arbitrary logic required). We already provide these pieces, so I think we are good to go.